src/librustc_typeck/astconv.rs

```diff
@@ -782,7 +780,8 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                     }
                 }
                 (GenericParamDefKind::Const, GenericArg::Const(ct)) => {
-                    self.ast_const_to_const(&ct.value, tcx.type_of(param.def_id)).into()
+                    let ct = tcx.hir().local_def_id(ct.value.hir_id);
+                    ty::Const::from_hir_anon_const(tcx, ct, tcx.type_of(param.def_id)).into()
```

Comment:
`tcx.type_of(param.def_id)` here may seem good but ultimately `tcx.type_of(ct_def_id)` has to work, otherwise the `AnonConst` can't be type-checked anyway.